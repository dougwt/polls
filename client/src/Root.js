import React from 'react';
import PropTypes from 'prop-types';
import { Provider } from 'react-redux';
import { createStore, applyMiddleware } from 'redux';
import { composeWithDevTools } from 'redux-devtools-extension';
import promiseMiddleware from 'redux-promise';
import reduxThunk from 'redux-thunk';
import reducers from './reducers';

const localAuth = JSON.parse(localStorage.getItem('auth') || null);

const Root = ({
  children,
  initialState = { auth: localAuth ? localAuth : false }
}) => {
  const store = createStore(
    reducers,
    initialState,
    composeWithDevTools(applyMiddleware(reduxThunk, promiseMiddleware))
  );

  return <Provider store={store}>{children}</Provider>;
};
Root.propTypes = {
  children: PropTypes.object.isRequired,
  initialState: PropTypes.object
};

export default Root;
