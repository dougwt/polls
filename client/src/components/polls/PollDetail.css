.container {
  margin-top: 4em;
  margin-bottom: 4em;
}

.card-title {
  /* font-style: italic; */
}

.choices {
  padding-top: 2em;
  padding-bottom: 2em;
}
