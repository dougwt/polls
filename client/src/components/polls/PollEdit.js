import React, { Component } from 'react';
import PropTypes from 'prop-types';
import { compose } from 'redux';
import { connect } from 'react-redux';
import { withRouter } from 'react-router-dom';
import requireAuth from '../requireAuth';
import PollForm from './PollForm';
import PollFormReview from './PollFormReview';
import * as actions from '../../actions';

export class PollEdit extends Component {
  constructor(props) {
    super(props);

    this.state = { showReview: false };
  }

  onCancel() {
    this.setState({ showReview: false });
  }

  onDelete() {
    this.props.deletePoll(this.props.pollId, () => {
      this.props.history.push('/polls');
    });
  }

  onSave(formValues) {
    const poll = {
      question: formValues.question,
      choices: formValues.choices.map((choice, i) => {
        return { text: formValues[`choice_${i + 1}`], votes: 0 };
      })
    };

    if (this.props.auth) {
      poll.owner = this.props.auth._id;
    }
    this.props.editPoll(this.props.pollId, poll, () => {
      this.props.history.push('/polls');
    });
  }

  onSubmit() {
    this.setState({ showReview: true });
  }

  renderContent() {
    if (this.state.showReview) {
      return (
        <PollFormReview
          onCancel={this.onCancel.bind(this)}
          onSave={this.onSave.bind(this)}
        />
      );
    }

    return (
      <PollForm
        onCancel={this.props.onCancel.bind(this)}
        onDelete={this.onDelete.bind(this)}
        onSubmit={this.onSubmit.bind(this)}
        initialValues={this.props.formValues}
      />
    );
  }

  render() {
    return this.renderContent();
  }
}
PollEdit.propTypes = {
  auth: PropTypes.oneOfType([PropTypes.object, PropTypes.bool]).isRequired,
  deletePoll: PropTypes.func.isRequired,
  editPoll: PropTypes.func.isRequired,
  formValues: PropTypes.object.isRequired,
  history: PropTypes.object.isRequired,
  onCancel: PropTypes.func.isRequired,
  polls: PropTypes.object,
  pollId: PropTypes.string
};

function mapStateToProps(state) {
  return {
    auth: state.auth
  };
}

export default compose(
  connect(mapStateToProps, actions),
  requireAuth,
  withRouter
)(PollEdit);
