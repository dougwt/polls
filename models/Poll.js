const mongoose = require('mongoose');
const { Schema } = mongoose;

const choiceSchema = require('./Choice');

const pollSchema = new Schema(
  {
    owner: {
      type: Schema.Types.ObjectId,
      ref: 'User'
    },
    question: String,
    choices: [choiceSchema],
    respondents: [
      {
        type: Schema.Types.ObjectId,
        ref: 'User'
      }
    ]
  },
  {
    usePushEach: true
  }
);

const Poll = mongoose.model('polls', pollSchema);

module.exports = Poll;
