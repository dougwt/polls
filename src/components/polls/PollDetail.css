.container {
  margin-top: 4em;
  margin-bottom: 4em;
}

.card-title {
  /* font-style: italic; */
}

.choices {
  padding-top: 2em;
  padding-bottom: 2em;
}

.choices label,
.choices .col {
  width: 100%;
}

.choices label:hover {
  color: #333;
  font-size: 1.1em;
}
