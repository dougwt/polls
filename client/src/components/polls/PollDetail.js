import React from 'react';
import PropTypes from 'prop-types';
import { connect } from 'react-redux';
import { Card, Input, Button, Row } from 'react-materialize';
import { votePoll } from '../../actions';
import { Pie } from 'react-chartjs-2';

import './PollDetail.css';

const defaultPoll = {
  choices: [0, 1, 2, 3, 4, 5],
  question: 'Who is your favorite Starfleet captain?',
  choice_1: 'Kirk (TOS)',
  choice_2: 'Picard (TNG)',
  choice_3: 'Sisko (DS9)',
  choice_4: 'Janeway (Voyager)',
  choice_5: 'Archer (Enterprise)',
  choice_6: 'Lorca (Discovery)'
};

const onVoteSubmit = (pollId, votePoll, disabled) => {
  const checked = document.querySelector('Input[name="choices"]:checked');
  const choiceId = checked ? checked.value : undefined;

  if (!disabled && pollId && choiceId) {
    votePoll(pollId, choiceId);
  }
};

export const PollDetail = ({
  auth,
  poll,
  formValues = defaultPoll,
  disabled = false,
  votePoll,
  waiting
}) => {
  return (
    <div className="container">
      <Card className="darken-1" title={formValues.question}>
        {poll && poll.respondents.includes(auth._id)
          ? renderResults(poll)
          : renderForm(poll, formValues, disabled, votePoll, waiting)}
      </Card>
    </div>
  );
};
PollDetail.propTypes = {
  auth: PropTypes.object,
  poll: PropTypes.object,
  formValues: PropTypes.object,
  disabled: PropTypes.bool,
  votePoll: PropTypes.func,
  waiting: PropTypes.bool
};

function renderForm(poll, formValues, disabled, votePoll, waiting) {
  return (
    <form
      className="poll"
      onSubmit={event => {
        event.preventDefault();
        onVoteSubmit(poll._id, votePoll, disabled);
      }}
    >
      {/* TODO: add spacing here via CSS */}

      <div className="choices">{renderChoices(poll, formValues)}</div>

      <Row className="center-align">
        {renderAsyncButton(waiting, disabled, event => {
          event.preventDefault();
          onVoteSubmit(poll._id, votePoll, disabled);
        })}
      </Row>
    </form>
  );
}

function renderResults(poll) {
  const data = { datasets: [{ data: [], backgroundColor: [] }], labels: [] };

  const colors = ['red', 'blue', 'green', 'yellow', 'orange', 'purple'];

  poll.choices.forEach((choice, i) => {
    data.datasets[0].data.push(choice.votes);
    data.datasets[0].backgroundColor.push(colors[i]);
    data.labels.push(choice.text);
  });

  const options = {
    legend: { position: 'bottom' }
  };

  return <Pie data={data} options={options} />;
}

function renderChoices(poll, values) {
  function renderChoice(key, value, label) {
    return (
      <Row key={key}>
        <Input name="choices" type="radio" value={value} label={label} />
      </Row>
    );
  }

  let choices = [];

  if (!poll) {
    // If poll is undefined it has not been saved, so use fake ids
    for (
      let choice = 1;
      choice <= (values.choices ? values.choices.length : 0);
      choice++
    ) {
      let field = `choice_${choice}`;
      if (values[field]) {
        choices.push(renderChoice(choice, choice.toString(), values[field]));
      }
    }
  } else {
    // If defined, we can use real ids from the existing poll
    poll.choices.forEach(choice => {
      choices.push(renderChoice(choice._id, choice._id, choice.text));
    });
  }

  return choices;
}

function renderAsyncButton({ waiting, disabled, onClick }) {
  if (waiting) {
    return (
      <Button className="teal" disabled={true}>
        Saving...
      </Button>
    );
  }

  return (
    <Button className="teal" disabled={disabled} onClick={onClick}>
      Vote
    </Button>
  );
}
renderAsyncButton.propTypes = {
  disabled: PropTypes.bool.isRequired,
  onClick: PropTypes.func.isRequired,
  waiting: PropTypes.bool.isRequired
};

function mapStateToProps(state) {
  return {
    auth: state.auth,
    error: state.poll.error,
    waiting: state.poll.waiting
  };
}

export default connect(mapStateToProps, { votePoll })(PollDetail);
