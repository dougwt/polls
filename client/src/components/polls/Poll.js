import React, { Component } from 'react';
import PropTypes from 'prop-types';
import { connect } from 'react-redux';
import { Button, Icon, Row, Section, Container } from 'react-materialize';
import PollDetail from './PollDetail';
import PollEdit from './PollEdit';
import { fetchPolls } from '../../actions';

function convertPollToFormValues(poll) {
  if (poll.length < 1) {
    poll = { question: '', choices: [] };
  }
  const formValues = {
    question: poll.question,
    choices: []
  };

  poll.choices &&
    poll.choices.forEach((choice, i) => {
      formValues.choices.push(i);
      formValues[`choice_${i + 1}`] = choice.text;
    });

  return formValues;
}

export class Poll extends Component {
  constructor(props) {
    super(props);

    this.state = { showEdit: false };
  }

  componentDidMount() {
    if (!this.props.fetched) {
      this.props.fetchPolls();
    }
  }

  renderContent() {
    const { pollId } = this.props.match.params;
    // TODO: this is inefficient! Switch to a map for faster lookups.
    let poll = this.props.polls.filter(poll => poll._id === pollId);
    if (poll.length > 0) {
      poll = poll[0];
    }

    if (this.state.showEdit) {
      return (
        <PollEdit poll={poll} formValues={convertPollToFormValues(poll)} />
      );
    }

    return (
      <div>
        <PollDetail poll={poll} formValues={convertPollToFormValues(poll)} />

        <Row className="right-align">
          <Button
            id="edit"
            className="btn"
            waves="light"
            node="a"
            flat={true}
            onClick={() => this.setState({ showEdit: true })}
          >
            Edit
            <Icon right>edit</Icon>
          </Button>
        </Row>
      </div>
    );
  }

  render() {
    return (
      <Section>
        <Container>
          <Row>
            <h3 className="header section-title">Poll</h3>
          </Row>
          {this.renderContent()}
        </Container>
      </Section>
    );
  }
}
Poll.propTypes = {
  polls: PropTypes.array,
  fetched: PropTypes.bool,
  match: PropTypes.object,
  fetchPolls: PropTypes.func
};

function mapStateToProps(state) {
  return {
    polls: state.poll.polls,
    fetched: state.poll.fetched
  };
}

export default connect(mapStateToProps, { fetchPolls })(Poll);
