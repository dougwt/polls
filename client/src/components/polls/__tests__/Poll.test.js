import React from 'react';
import { shallow, mount } from 'enzyme';
import { Poll } from '../Poll';
import PollDetail from '../PollDetail';
import PollEdit from '../PollEdit';
import Root from '../../../Root';

describe('Poll', () => {
  let props;
  let wrapper;

  beforeEach(() => {
    props = {
      polls: [],
      fetched: false,
      match: { params: { pollId: 1 } },
      fetchPolls: jest.fn()
    };

    wrapper = shallow(<Poll {...props} />);
  });

  it('renders properly', () => {
    expect(wrapper).toMatchSnapshot();
  });

  describe('when the page is initially loaded', () => {
    it('contains the proper component state', () => {
      expect(wrapper.state('showEdit')).toEqual(false);
    });

    it('shows the PollDetail component', () => {
      expect(wrapper.find(PollDetail).length).toEqual(1);
    });

    it('calls the fetchPolls action creator', () => {
      wrapper = mount(
        <Root>
          <Poll {...props} />
        </Root>
      );

      expect(props.fetchPolls).toHaveBeenCalled();
    });
  });

  describe('when the edit button is clicked', () => {
    beforeEach(() => {
      wrapper.find('#edit').simulate('click');
      wrapper.update();
    });

    it('updates the component state', () => {
      expect(wrapper.state('showEdit')).toEqual(true);
    });

    it('shows the PollEdit component', () => {
      expect(wrapper.find(PollEdit).length).toEqual(1);
    });

    it('does not show the PollDetail component', () => {
      expect(wrapper.find(PollDetail).length).toEqual(0);
    });
  });
});
