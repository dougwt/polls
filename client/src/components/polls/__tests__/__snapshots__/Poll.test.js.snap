// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Poll renders properly 1`] = `
ShallowWrapper {
  Symbol(enzyme.__root__): [Circular],
  Symbol(enzyme.__unrendered__): <Poll
    fetchPolls={[MockFunction]}
    fetched={false}
    match={
      Object {
        "params": Object {
          "pollId": 1,
        },
      }
    }
    polls={Object {}}
  />,
  Symbol(enzyme.__renderer__): Object {
    "batchedUpdates": [Function],
    "getNode": [Function],
    "render": [Function],
    "simulateError": [Function],
    "simulateEvent": [Function],
    "unmount": [Function],
  },
  Symbol(enzyme.__node__): Object {
    "instance": null,
    "key": undefined,
    "nodeType": "class",
    "props": Object {
      "children": <Container>
        <Row>
          <h3
            className="header section-title"
          >
            Poll
          </h3>
        </Row>
        Loading...
      </Container>,
      "node": "div",
    },
    "ref": null,
    "rendered": Object {
      "instance": null,
      "key": undefined,
      "nodeType": "function",
      "props": Object {
        "children": Array [
          <Row>
            <h3
              className="header section-title"
            >
              Poll
            </h3>
          </Row>,
          "Loading...",
        ],
      },
      "ref": null,
      "rendered": Array [
        Object {
          "instance": null,
          "key": undefined,
          "nodeType": "function",
          "props": Object {
            "children": <h3
              className="header section-title"
            >
              Poll
            </h3>,
          },
          "ref": null,
          "rendered": Object {
            "instance": null,
            "key": undefined,
            "nodeType": "host",
            "props": Object {
              "children": "Poll",
              "className": "header section-title",
            },
            "ref": null,
            "rendered": "Poll",
            "type": "h3",
          },
          "type": [Function],
        },
        "Loading...",
      ],
      "type": [Function],
    },
    "type": [Function],
  },
  Symbol(enzyme.__nodes__): Array [
    Object {
      "instance": null,
      "key": undefined,
      "nodeType": "class",
      "props": Object {
        "children": <Container>
          <Row>
            <h3
              className="header section-title"
            >
              Poll
            </h3>
          </Row>
          Loading...
        </Container>,
        "node": "div",
      },
      "ref": null,
      "rendered": Object {
        "instance": null,
        "key": undefined,
        "nodeType": "function",
        "props": Object {
          "children": Array [
            <Row>
              <h3
                className="header section-title"
              >
                Poll
              </h3>
            </Row>,
            "Loading...",
          ],
        },
        "ref": null,
        "rendered": Array [
          Object {
            "instance": null,
            "key": undefined,
            "nodeType": "function",
            "props": Object {
              "children": <h3
                className="header section-title"
              >
                Poll
              </h3>,
            },
            "ref": null,
            "rendered": Object {
              "instance": null,
              "key": undefined,
              "nodeType": "host",
              "props": Object {
                "children": "Poll",
                "className": "header section-title",
              },
              "ref": null,
              "rendered": "Poll",
              "type": "h3",
            },
            "type": [Function],
          },
          "Loading...",
        ],
        "type": [Function],
      },
      "type": [Function],
    },
  ],
  Symbol(enzyme.__options__): Object {
    "adapter": ReactSixteenAdapter {
      "options": Object {
        "enableComponentDidUpdateOnSetState": true,
        "lifecycles": Object {
          "componentDidUpdate": Object {
            "onSetState": true,
          },
          "getDerivedStateFromProps": true,
          "getSnapshotBeforeUpdate": true,
          "setState": Object {
            "skipsComponentDidUpdateOnNullish": true,
          },
        },
      },
    },
    "disableLifecycleMethods": true,
  },
}
`;
